### ISSUE TITLE
Multi-GPU communication outputs

### ISSUE THREAD

**Issue by SimlaBurcu** [NVIDIA/PyProf#33]
Hi everyone,
I’m trying to profile a distributed data parallel training of a deep learning model running on 2 GPUs. The model is implemented in PyTorch, and uses PyTorch’s NCCL back end for GPU-to-GPU communication.
I am using PyProf on Linux terminal and the outputs I’m getting for the kernel times are reasonable for computation.
But for communication, I am getting strange results:
First of all, PyProf behaves PyTorch’s nccl all_reduce/broadcast calls as kernels (namely ncclAllReduceRingLLKernel_sum_f32).

1. Since I’m profiling both GPUs, I can compare the times spent for communication. There is a big difference in two GPUs for nccl kernels (10x to 1000x). When I measure nccl time with “cuda.Event’s elapsed_time” two GPUs seem to spend the same time and the bandwidth (calculated with that measurement) is reasonable.
- Which measurement is the correct one?
- Why PyProf can’t measure communication time?
- Is cuda.Event good for profiling?

2. Moreover the GPU that has larger time changes between training iterations. For example if GPU1 spends 1 sec for allreduce, and GPU2 spends 0.1 sec, after 3 iterations, GPU2 starts spending 1 sec while GPU1 spends 0.1
3. Last question: When I change back end to gloo, it doesn’t give any output for communication. I wonder the reason for this.

**Comment by adityaiitb** [comment]
hi,

+ @mkolod 

Note that NCCL calls are GPU kernels which do communication + computation (accumulation of gradients) and so you will see them on the compute stream. You can verify this with vanilla NVprof / Nsight Systems as well. PyProf uses NVprof / Nsight and reports everything on the compute streams.

1. In your setup are you overlapping NCCL calls with other GPU kernels? If yes, they will compete for GPU resources (SM, DRAM bandwidth) and so the time spent will vary. Do you see NCCL calls in the middle of backprop OR are all NCCL calls at the end of back prop (before optimizer)? If its the former it means, your NCCL calls are overlapping with other compute kernels. For the sake of experimentation, you can change the buffer size of NCCL and force all communication/gradient reduction to happen at the end of back prop. Not sure how to do it though.

I have not used `cuda.Event` recently. When you use `cuda.Event` does that force any sort of synchronization? Because that would explain the deterministic time.

2. Non determinism probably related to NCCL competing with other GPU kernels.

3. I have never used gloo. I will have to use and read up on that.

**Comment by SimlaBurcu** [comment]
Thank you @adityaiitb  for your quick response.

1. I am using an implemented model, I'm not sure how to check whether NCCL is overlapping or not? 
* But based on the results, it seems like they are overlapping and compete for the resources like you said. Yes, they are in the middle of backprop. Moreover the system also has model parallelism, and the inconsistency in nccl times exists there too.
* In order to measure time with `cuda.Event` you need to call `torch.cuda.synchronize()`, so yes. 
* Also, I'm using the terminal and get the csv outputs. So what I have is the times for each kernel. I know that nvprof won't have the kernels execute concurrently. If I want to see the overall time spent for forward prop for example, is it possible to get accurate results by summing all the kernels?

**Comment by adityaiitb** [comment]
You can open the log file in NVVP / Nsight and check if NCCL kernels overlap with other GPU kernels.

Correction: Note that NVprof does not alter the concurrency (parallel execution on multiple CUDA streams) specified in the program. The reason you see *almost* all kernels in a single stream is because PyTorch launches all kernels in a single stream (one stream for forward kernels, another stream for backward kernels). NCCL kernels are *probably* launched on a separate stream and hence you should notice concurrency. You can use NVVP / Nsight to verify this visually. You can also use PyProf to dump the stream_id for every kernel.

Also note that NVprof / PyProf might introduce CPU overheads (manifested as slightly larger gaps between kernels) but it does not impact the GPU kernel time itself. Since, all forward kernels are launched in a single stream, you can simply add the "GPU kernel times" to get the time spent on the GPU in fprop. However, if you want to measure the wall clock time spent on fprop, I would suggest to turn off the profiler and do something like

```
device or stream synchronize
start = time.time()
fprop kernels
device or stream synchronize # wait for all fprop kernels to finish
duration = time.time() - start
```
If your GPU kernels are big (say > 100 us) then wall clock time ~= GPU kernel time.

**Comment by debermudez** [comment]
Closing due to lack of activity
